### PR TITLE
[expo-notifications] Move stores to service package and migrate to Kotlin

### PR DESCRIPTION
**Pull request by sjchmiela**
# Why

We want to Kotlinify our database + stores belong to `.service` package

# How

- drag and dropped classes
- transformed to Kotlin automatically
- rewrote some of the functions to make them more Kotliny

# Test Plan

`test-suite` passes all relevant tests.

### CODE REVIEW

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/SharedPreferencesNotificationCategoriesStore.kt [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+import java.io.ByteArrayOutputStream
+import java.io.IOException
+import java.io.InvalidClassException
+import java.io.ObjectInputStream
+import java.io.ObjectOutputStream
+
+/**
+ * A fairly straightforward [SharedPreferences] wrapper to be used by [NotificationSchedulingHelper].
+ * Saves and reads notification category information (identifiers, actions, and options) to and from persistent storage.
+ *
```
Comment (bbarthec) [inline packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/SharedPreferencesNotificationCategoriesStore.kt:17]:
```suggestion
```

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/SharedPreferencesNotificationCategoriesStore.kt [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    deserializeNotificationCategory(
+      sharedPreferences.getString(
+        preferencesNotificationCategoryKey(identifier),
+        null
+      )
+    )
+
+  /**
+   * Fetches all categories, ignoring invalid ones.
+   *
```
Comment (bbarthec) [inline packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/SharedPreferencesNotificationCategoriesStore.kt:50]:
```suggestion
```

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/SharedPreferencesNotificationCategoriesStore.kt [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+          try {
+            return@mapNotNull deserializeNotificationCategory(value as String?)
+          } catch (e: ClassNotFoundException) {
+            // do nothing
+          } catch (e: IOException) {
+            // do nothing
+          }
+        }
+        null
+      }
```
Comment (bbarthec) [inline packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/SharedPreferencesNotificationCategoriesStore.kt:70]: I believe that's the perfect moment for adding this one: https://developer.android.com/kotlin/ktx

Comment (bbarthec) [reply packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/SharedPreferencesNotificationCategoriesStore.kt:70]: And what would you say for writing it this way:
```suggestion
      sharedPreferences.all
        .filter(key -> key.startsWith(NOTIFICATION_CATEGORY_KEY_PREFIX))
        .mapNotNull { (key, value) -> 
          return try {
            deserializeNotificationCategory(value as String?)
          } catch (e: ClassNotFoundException) {
            null
          } catch (e: IOException) {
            null
          }
        }
```
less indents & `if`s, better readability 😉

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/SharedPreferencesNotificationsStore.kt [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+import java.io.ByteArrayOutputStream
+import java.io.IOException
+import java.io.InvalidClassException
+import java.io.ObjectInputStream
+import java.io.ObjectOutputStream
+
+/**
+ * A fairly straightforward [SharedPreferences] wrapper to be used by [NotificationSchedulingHelper].
+ * Saves and reads notifications (identifiers, requests and triggers) to and from the persistent storage.
+ *
```
Comment (bbarthec) [inline packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/SharedPreferencesNotificationsStore.kt:17]:
```suggestion
```

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/SharedPreferencesNotificationsStore.kt [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+      sharedPreferences.getString(
+        preferencesNotificationRequestKey(identifier),
+        null
+      )
+    )
+
+  /**
+   * Fetches all scheduled notifications, ignoring invalid ones.
+   *
+   *
```
Comment (bbarthec) [inline packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/SharedPreferencesNotificationsStore.kt:51]:
```suggestion
```

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/SharedPreferencesNotificationsStore.kt [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        try {
+          if (key.startsWith(NOTIFICATION_REQUEST_KEY_PREFIX)) {
+            return@mapNotNull deserializeNotificationRequest(value as String?)
+          }
+        } catch (e: ClassNotFoundException) {
+          // do nothing
+        } catch (e: IOException) {
+          // do nothing
+        }
+        null
```
Comment (bbarthec) [inline packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/SharedPreferencesNotificationsStore.kt:69]: Same suggestion as in other file

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/SharedPreferencesNotificationsStore.kt [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+   * @return Deserialized trigger
+   * @throws IOException            Thrown if there is an error while reading trigger from String
+   * @throws ClassNotFoundException Thrown if the deserialization failes due to class not being found.
+   * @throws InvalidClassException  Thrown if the trigger is of invalid class.
+   */
+  @Throws(IOException::class, ClassNotFoundException::class, InvalidClassException::class)
+  private fun deserializeNotificationRequest(trigger: String?): NotificationRequest =
+    Base64.decode(trigger, Base64.NO_WRAP).let {
+      ByteArrayInputStream(it).use { byteArrayInputStream ->
+        ObjectInputStream(byteArrayInputStream).use { ois ->
```
Comment (bbarthec) [inline packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/SharedPreferencesNotificationsStore.kt:151]: It seems similar to the deserialisation from the other file? 🤔 
Maybe it's ok to ship some helper method for this?

**Review by bbarthec — COMMENTED** [review]
LGTM, but I've left some comments 😉

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/SharedPreferencesNotificationsStore.kt (hunk 7 above)
Comment (sjchmiela) [reply packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/SharedPreferencesNotificationsStore.kt]: I'm lovin it! https://github.com/expo/expo/pull/10765/commits/d5d0b24a608e4f04fd0689882e3d0f9713a9ee72

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/Base64Serialization.kt [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+inline fun <reified T> String.asBase64EncodedObject(): T =
+  ByteArrayInputStream(
+    Base64.decode(this, Base64.NO_WRAP)
+  ).use { byteArrayInputStream ->
+    ObjectInputStream(byteArrayInputStream).use { ois ->
+      val o = ois.readObject()
+      if (o is T) {
+        return o
+      }
+      throw InvalidClassException("Expected serialized notification category to be an instance of NotificationCategory. Found: $o")
```
Comment (bbarthec) [inline packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/Base64Serialization.kt:31]:
```suggestion
      throw InvalidClassException("Expected serialized $getObjectType to be an instance of $getTheGenericClass(). Found: $o")
```

**Review by bbarthec — APPROVED** [review]
Yeah, looks solid 👍 
Only if you can adjust the error message as the current seems to be misleading 🤔

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/Base64Serialization.kt (hunk 8 above)
Comment (sjchmiela) [reply packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/Base64Serialization.kt:31]: Thank you for catching that! ❤️

Comment (sjchmiela) [reply packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/Base64Serialization.kt]: I went for `Expected serialized object to be an instance of ${T::class.java}. Found: $o`.

**Pull request merged**